pkg/hintrunner/zero/zerohint_math.go

```diff
@@ -590,6 +590,106 @@ func createSplitFeltHinter(resolver hintReferenceResolver) (hinter.Hinter, error
 	return newSplitFeltHint(low, high, value), nil
 }
 
+func newSignedDivRemHint(value, div, bound, r, biased_q hinter.ResOperander) hinter.Hinter {
+	return &GenericZeroHinter{
+		Name: "SignedDivRem",
+		Op: func(vm *VM.VirtualMachine, _ *hinter.HintRunnerContext) error {
+			//> from starkware.cairo.common.math_utils import as_int, assert_integer
+			// assert_integer(ids.div)
+			// assert 0 < ids.div <= PRIME // range_check_builtin.bound, f'div={hex(ids.div)} is out of the valid range.'
+			// assert_integer(ids.bound)
+			// assert ids.bound <= range_check_builtin.bound // 2, f'bound={hex(ids.bound)} is out of the valid range.'
+			// int_value = as_int(ids.value, PRIME)
+			// q, ids.r = divmod(int_value, ids.div)
+			// assert -ids.bound <= q < ids.bound, f'{int_value} / {ids.div} = {q} is out of the range [{-ids.bound}, {ids.bound}).'
+			// ids.biased_q = q + ids.bound
+
+			//> assert_integer(ids.div)
+			divFelt, err := hinter.ResolveAsFelt(vm, div)
+			if err != nil {
+				return err
+			}
+			//> assert 0 < ids.div <= PRIME // range_check_builtin.bound, f'div={hex(ids.div)} is out of the valid range.'
+			upperBoundBig := new(big.Int).Div(fp.Modulus(), new(big.Int).Lsh(big.NewInt(1), 128))
+			upperBound := new(fp.Element).SetBigInt(upperBoundBig)
+			if !utils.FeltLe(divFelt, upperBound) {
+				return fmt.Errorf("div=%v is out of the valid range", divFelt)
+			}
+
+			//> assert_integer(ids.bound)
+			boundFelt, err := hinter.ResolveAsFelt(vm, bound)
+			if err != nil {
+				return err
+			}
+			//> assert ids.bound <= range_check_builtin.bound // 2, f'bound={hex(ids.bound)} is out of the valid range.'
+			felt127 := new(fp.Element).SetBigInt(new(big.Int).Lsh(big.NewInt(1), 127))
+			if !utils.FeltLe(boundFelt, felt127) {
+				return fmt.Errorf("bound=%v is out of the valid range", boundFelt)
+			}
+			//> int_value = as_int(ids.value, PRIME)
+			valueFelt, err := hinter.ResolveAsFelt(vm, value)
+			if err != nil {
+				return err
+			}
+
+			//> q, ids.r = divmod(int_value, ids.div)
+			var valueBig, divBig, boundBig big.Int
+			valueFelt.BigInt(&valueBig)
+			divFelt.BigInt(&divBig)
+			boundFelt.BigInt(&boundBig)
+			qBig, rBig := new(big.Int).DivMod(&valueBig, &divBig, new(big.Int))
+			qFelt, rFelt := new(fp.Element).SetBigInt(qBig), new(fp.Element).SetBigInt(rBig)
+			//> assert -ids.bound <= q < ids.bound, f'{int_value} / {ids.div} = {q} is out of the range [{-ids.bound}, {ids.bound}).'
+			if new(big.Int).Abs(&boundBig).Cmp(new(big.Int).Abs(qBig)) == -1 {
+				return fmt.Errorf("%v / %v = %v is out of the range [-%v, %v]", valueFelt, divFelt, qBig, boundFelt, boundFelt)
+			}
```

Comment:
Isn't this only raising an error if `ids.bound < q`? Shouldn't there be an error for `q == ids.bound` too? 🤔

Reply:
You are right, it should compare ids.bound <= q

Reply:
I don't think using just one comparison will solve this as the range limits are inclusive and exclusive respectively. So you'll have to check for the equality in just one case.
I see two choices:
- You can do the checks separately as in `-ids.bound <= q && q < ids.bound`
- You can continue to use `Abs()` method, but you'll need to add another condition as in `Abs(q) < Abs(ids.bound) || q == -ids.bound`

Reply:
I dont think the first equation will work in the field arithmetics, as -ids.bound (PRIME - ids.bound) is greater than ids.bound, when ids.bound is greater then PRIME // 2. The range would have to look more like q < ids.bound || q >= -ids.bound

Reply:
Added the proper check, handling that using big ints, not felts